### DISCUSSION TITLE
The need for a standardized OpenTelemetry Query API

### DISCUSSION THREAD

**Discussion by itaykat** [teletrace/teletrace#41] [Ideas]
Hi all, 

I had encountered this [discussion](https://github.com/open-telemetry/oteps/issues/193#issue-1075332027) and wanted to set the stage for a broader discussion regarding the need of a standardized OpenTelemetry backend (Query API... you name it) specifically for the query use cases. 

So in order for all of us and for future collaborators, in this discussion I'll share some of the needs and requirements described in the issue mentioned above.

I give the full credit to [lucasponce](https://github.com/lucasponce) for being so descriptive and I will summarize some of the issues he was raising along with some more needs and feedbacks from the rest of the OpenTelemetry community.

As we all know the OpenTelemetry project focused (as mentioned in their main site) in [_instrument, generate, collect, and export telemetry data (metrics, logs, and traces)_](https://opentelemetry.io/)

The OpenTelemetry project is currently not focused on storing and querying telemetry data and this come as a need from multiple use-cases.

Feel free to comment and share ideas. Cheers!

RELATED MIRO DIAGRAM: https://miro.com/app/board/uXjVPaTF5dc=/

**Comment by itaykat** [comment]
Following [lucasponce](https://github.com/lucasponce) [discussion on the OpenTelemetry repo:](https://github.com/open-telemetry/oteps/issues/193)

The need for an agnostic OpenTelemetry backend specification to consume traces stored in any tracing platform (Jaeger, Tempo, Zipkin) was raised multiple times in different use case context.

Today, applications can use opentelemetry sdk to ingest traces and spans in a common format. 
There are some platforms that consume telemetry data (metrics, traces and logs) to offer to the user added value on top of their core data (i.e Service Mesh platforms like [Kiali](https://kiali.io/).

One of the request from users who uses those platforms and developers who build it is the possibility to change the tracing platform i.e. https://github.com/kiali/kiali/issues/4278 but unfortunately if you want to query traces for a specific platform you'd require some technical dependency on that platform (in the Kiali case, this one is Jaeger).

There are proto definitions for a gRPC service but it's not obvious to change from one platform or other.

So, instead of waiting for specific implementations (Jaeger, Zipkin, Tempo, others) to implement a standardized common API for backend queries, it would be nice to have a single dependency (lets call it opentelemetry-backend) that can enable to query one or another backend in a common format.

Ideally users should have the ability to easily change from one OpenTelemetry backend to another (from the backend user perspective).

So main entities should be:
Instrumented App -> OTel Collector -> OTel Storage <- OTel Backend (Query API)<- 3rd Party Platform

As for now, the OpenTelemetry main project goal was to set free the instrumented application to choose their own telemetry ingesters (the platform where telemetry data is ingested and stored, i.e Jaeger, Zipkin, Tempo.

The goal for such backend is to have a unified way to query the storage of the telemetry data (in the first step, the tracing data).

OpenTelemetry standard is focused on a standard for instrumented applications for collectors, but there is a gap (or no standard) in a common API to be available to query the stored data.

I think this effort could be interesting for the OpenTelemetry community and it would be nice if it's fostered under the OpenTelemetry umbrella allowing users to participate more actively.

Any feedback would be welcome

Related discussion: https://github.com/open-telemetry/oteps/issues/193

**Comment by galbash** [comment]
This is indeed a great point (@osherv ^). I think that in order to contribute here we need to also acknowledge that there are efforts already going on in the community (TraceQL by Grafana), so we need to decide if we want to collaborate or compete with them. CC @razlo for business context

**Reply by itaykat** [reply]
TraceQL is a query syntax based on top of a proprietary non-standardized query api.

This is highly related as an example of a need to have a dedicated query language for the end user to use the query api in a more continent way.

@tvaintrob FYI. We can open a dedicated discussion for that topic.

**Comment by itaykat** [comment]
Another related discussion with a guy from Tempo: https://github.com/grafana/tempo/issues/1139

**Comment by osherv** [comment]
That's a very interesting point. I think that this will be very hard to convince the community to work on. Maybe we can try to take over his [issue](https://github.com/open-telemetry/oteps/issues/193) and see if we can push this.
Also, I think that since we want to develop anyway an open source OTel backend maybe we can start to develop it and then try to standard it in the community and all other Tracing tools.

**Reply by itaykat** [reply]
So the question is, does the OTLP specifications is enough for defining such an API? 
i.e. There is no definition for a Trace request.

**Reply by osherv** [reply]
I don't think so. We can try with a new OTEP

**Reply by ronyis** [reply]
OTLP specs could probably be a basis for many of the data types, but it definitely something else (probably should be separated as well)

**Reply by itaykat** [reply]
Quick note, Tempo has adapted the OpenTelemetry Line Protocol (OTLP) as data format (for storage).

**Comment by itaykat** [comment]
Please see a list of related discussions from the OpenTelemetry related Slack channels, feel free to add more related discussions to that page: https://www.notion.so/epsagon/Observability-Backend-Discussions-6230558e91de41078259486ad6d43f9d

**Reply by galbash** [reply]
Side note, we should probably find a public place for this. How does other OS organizations organize knowledge? Wiki?

**Reply by itaykat** [reply]
These discussions are only for reference, I will restructure and summarize them to be more coherent.

**Comment by ronyis** [comment]
Great point. Some thoughts:

* I think that having a shared API for querying all telemetry data could fit well into OTel specifications. Probably a gRPC API, and potentially including API libs in different runtimes.

* Next step for OTel could be to have a contrib repo including implementations of this API as a proxy over different data storages / tools (e.g. Tempo, Elasticsearch, Prometheus, etc.). A new OTel-based tool like we plan could just use this API natively.

* One of the challenges I see, is different tools having different capabilities and APIs. For example, one tool support searching by arbitrary attributes, aggregations, correlations between telemetry types, etc. Probably some parts of the API should be mandatory, and some optional.

**Reply by roybh1** [reply]
I think it could be beneficial to standardize each capability separately and allow mixing-and-matching different backends (searching arbitrary attributes with one backend, query type correlations with another etc)

**Reply by itaykat** [reply]
> I think it could be beneficial to standardize each capability separately and allow mixing-and-matching different backends (searching arbitrary attributes with one backend, query type correlations with another etc)

What do you mean by mix-match different be? Can you elaborate?

**Reply by galbash** [reply]
for example prometheus for metrics and elastic for logs

**Reply by tvaintrob** [reply]
I think it would be very beneficial separating the definition of the different capabilities, but in a different approach then what @roybh1 and @galbash suggested, I think what Microsoft did with their LSP specification is a better approach here, basically the defined all the capabilities a tool should have, and each backend can choose to implement all of them or a subset, then when initializing the backend will relay the capabilities that are supported and each client implementation can act accordingly 

[lsp capabilities section](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#capabilities)

**Reply by roybh1** [reply]
@itaykat @galbash @tvaintrob  I meant mixing and matching specifically tracing capabilities as @ronyis said - searching arbitrary attributes, span aggregation, correlations between telemetries etc (and not logs and metrics, which should also be supported). If we take the LSP example, the traces backend API will define those capabilities separately so the user should be able to use multiple backends simultaneously - for example elastic for full text search / tag search and timescale for graph lookups. Some backends are better than others depending on the use case.

This would also work well with the current OTel flow:
* OTel collector natively supports exporting the same data to multiple exporters simultaneously
* Multiple exporters may ingest to multiple backends
* Backend query API will allow querying those multiple backends to optimize different capabilities

**Reply by galbash** [reply]
@tvaintrob generally sounds good, I still think of course we can build the language incrementally and start with the traces, but sounds good.

**Comment by ronyis** [comment]
I think that the next step could be creating an actual OTEP about it (current [discussion](https://github.com/open-telemetry/oteps/issues/193#issue-1075332027) is still just an issue).

The OTEP should already include an example of how this API will look.

**Comment by itaykat** [comment]
Reviewing the discussion again, this need that comes from multiple users can be interpreted in two ways of implementation:

1.   An extensible query layer that comes as part of a fully featured observability backend, and can support multiple databases for storing traces. This way the user can have a free will to choose his own database which he is already maintaining.

2. An extensible query layer that wraps existing open-source observability backends (OB) (Jaeger, Tempo, etc.), this is the main issue that is being described in the discussion above.

@ronyis @tvaintrob @haimrait we should decide between those two.

**Reply by itaykat** [reply]
@haimrait I opened a placeholder for that issue #64

**Reply by itaykat** [reply]
So I've discussed with @haimrait about the two ways mentioned above and decided that the 2nd is a subset of the 1st (as we can relate to an OB as a specific datasource and build a dedicated plugin for it)

**Comment by itaykat** [comment]
Suggested draft for layered architecture:

![Screen Shot 2022-09-04 at 15 54 06](https://user-images.githubusercontent.com/67950726/188314523-3051fdf7-a3b4-48cf-ae4f-6afc30495ab4.png)

**Reply by ronyis** [reply]
I think that at this point, we should push for adding the specification layer (and maybe also the service layer) to OpenTelemetry.
We could earn much by being the first tool that natively supports it.

**Reply by itaykat** [reply]
I agree. @haimrait @osherv any input from your side? Otherwise I'll open an issue for me to create the discussed OTEP.

**Comment by itaykat** [comment]
An ongoing architecture design for the query layer:
https://miro.com/app/board/uXjVPaTF5dc=/

**Comment by itaykat** [comment]
@razlo @haimrait please take a look at the last messages in this [discussion](https://github.com/open-telemetry/oteps/issues/193). 
Copy paste last message:

>> In a nutshell, there is a need to query telemetry data stored in backends.
>
>This is not very well aligned with the current vision of OpenTelemetry. Otel in the past mostly stayed away from backend functionality. See https://opentelemetry.io/docs/ where we say:
>
>>OpenTelemetry, also known as OTel for short, is a vendor-neutral open-source [Observability](https://opentelemetry.io/docs/concepts/observability-primer/#what-is-observability) framework for instrumenting, generating, collecting, and exporting telemetry data such as [traces](https://opentelemetry.io/docs/concepts/observability-primer/#distributed-traces), [metrics](https://opentelemetry.io/docs/concepts/observability-primer/#reliability--metrics), [logs](https://opentelemetry.io/docs/concepts/observability-primer/#logs).
>
>See how it stops at "exporting". With some small exceptions we consider our job done once the telemetry is delivered to the backend.
>
>That said, I do understand that how it can valuable to have a telemetry query standard. If there is enough interest in this then potentially Otel may consider working on it. This is a major initiative though and initiatives like this typically are started by folks who are interested in the topic and form a workgroup and propose it as a project (which may or may not be accepted by Otel). See more about relevant processes [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/project-management.md) and [here](https://github.com/open-telemetry/oteps#evolving-opentelemetry-at-the-speed-of-markdown). Step 1 would be to find sufficiently large group of people with relevant experience willing to spend significant time on this initiative.

**Step 1 would be to find sufficiently large group of people with relevant experience willing to spend significant time on this initiative.**

We will talk about this initiative on tomorrow meeting.

**Comment by itaykat** [comment]
Trying to chart some of the input we have here,[ Miro link](https://miro.com/app/board/uXjVPaTF5dc=/):
<img width="680" alt="Screen Shot 2022-09-07 at 16 38 52" src="https://user-images.githubusercontent.com/67950726/188892609-262b2144-31a4-4b72-be4e-8cb2ad35c2f3.png">

**Comment by itaykat** [comment]
New OTEP was created, still draft: https://github.com/epsagon/oteps-drafts/pull/2

**Comment by galbash** [comment]
@itaykat check out @schultyy project: https://github.com/epsagon/trace-insights . This can be something we can advocate the API with - whoever supports the traces API we can promote the plugin to support as well

**Reply by itaykat** [reply]
Great example! Was not familiar with that project.
Actually all telemetry profiling related IDE plugins are relevant. We need to find more of them.

@schultyy feel free to have your take on that OTEP: https://github.com/open-telemetry/oteps/pull/218
